pkg/controller/fileobserver/observer_linux.go

```diff
@@ -0,0 +1,116 @@
+// +build linux
+
+package fileobserver
+
+import (
+	"path/filepath"
+
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/golang/glog"
+	"github.com/sigma/go-inotify"
+)
+
+type linuxObserver struct {
+	watcher  *inotify.Watcher
+	queue    workqueue.Interface
+	reactors map[string][]reactorFn
+}
+
+func init() {
+	NewObserver = newLinuxObserver
+}
+
+func newLinuxObserver() (Observer, error) {
+	inotifyWatcher, err := inotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	return &linuxObserver{
+		watcher:  inotifyWatcher,
+		reactors: map[string][]reactorFn{},
+		queue:    workqueue.New(),
+	}, nil
+}
+
+func (o *linuxObserver) AddReactor(reactor reactorFn, filenames ...string) Observer {
+	for _, f := range filenames {
+		glog.V(5).Infof("Adding reactor for file %q", f)
+		o.watcher.AddWatch(filepath.Dir(f), inotify.IN_MODIFY|inotify.IN_CREATE|inotify.IN_DELETE)
+		o.reactors[f] = append(o.reactors[f], reactor)
+	}
+	return o
+}
+
+type reactionImpl struct {
+	action   ObservedFileAction
+	filename string
+}
+
+func (o *linuxObserver) processQueueReactions() {
+	for {
+		obj, shutdown := o.queue.Get()
+		if shutdown {
+			return
+		}
+		r := obj.(*reactionImpl)
+		reactorFns := o.reactors[r.filename]
+		for i := range reactorFns {
+			if err := reactorFns[i](r.filename, r.action); err != nil {
+				glog.Errorf("Reactor for %q failed: %v", obj.(string), err)
+			}
+		}
+		o.queue.Done(obj)
+	}
+}
+
+func maskToActionType(mask uint32) ObservedFileAction {
+	switch mask {
+	case inotify.IN_MODIFY:
+		return ObservedFileModified
+	case inotify.IN_DELETE:
+		return ObservedFileDeleted
+	case inotify.IN_CREATE:
+		return ObservedFileCreated
+	default:
+		// Should never happen as we only handle the three events above.
+		return ObservedFileUnknownAction
+	}
+}
+
+func (o *linuxObserver) reportWatcherErrors() {
+	for err := range o.watcher.Error {
+		glog.Error(err)
+	}
+}
+
+func (o *linuxObserver) processWatchEvents() {
+	for event := range o.watcher.Event {
+		if _, hasReactor := o.reactors[event.Name]; !hasReactor {
+			continue
+		}
+		glog.Infof("Observed change %s", event.String())
+		o.queue.Add(&reactionImpl{action: maskToActionType(event.Mask), filename: event.Name})
+	}
+}
+
+func (o *linuxObserver) Run(stopCh <-chan struct{}) {
+	glog.Info("Starting file observer")
+
+	// Process all errors from the watcher in parallel
+	go o.reportWatcherErrors()
```

Comment:
go funcs need stop channels.  Also, you need to catch panics `utilruntime.somethin`